README.md
<!-- PROJECT LOGO -->
<br />
<p align="center">
  <a href="https://github.com/DominoVagrant/demo-v2-task-based/">
    <img src="https://startcloud.com/assets/logo-big.jpg" alt="Logo" width="200" height="100">
  </a>

  <h3 align="center">Domino Vagrant Build</h3>
  <h3> This has been archived! This was a demo for Super.Human.Installer, Please see https://github.com/STARTcloud/hcl_domino_standalone_provisioner</h3>
  

  <p align="center">
    An README to jumpstart your build of the Domino Development
    <br />
    <a href="https://github.com/DominoVagrant/demo-v2-task-based/"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/DominoVagrant/demo-v2-task-based/">View Demo</a>
    ·
    <a href="https://github.com/DominoVagrant/demo-v2-task-based/issues">Report Bug</a>
    ·
    <a href="https://github.com/DominoVagrant/demo-v2-task-based/issues">Request Feature</a>
  </p>
</p>

<!-- TABLE OF CONTENTS -->
## Table of Contents

* [About the Project](#dominovagrant)
  * [Built With](#built-with)
* [Getting Started](#getting-started)
  * [Prerequisites](#prerequisites)
  * [Installation](#installation)
    * [Mac OS X](https://github.com/DominoVagrant/demo-v2-task-based/blob/master/MacMojaveReadme.md) -- Quick Start
    * [Windows](https://github.com/DominoVagrant/demo-v2-task-based/blob/master/Win10ReadMe.md) -- Quick Start
* [Deployment](#deployment)
  * [Cloning](#cloning-the-repo-locally)
  * [Overview](#configuring-the-environment)
  * [Variables](#commonly-changed-parameters)
  * [Source Files](#source-files)
* [Initialization](#starting-the-vm)
  * [Access Methods](#accessing-the-domino-server)
    * [Web](#web-interface)
    * [Notes Client](#access-from-notes-client)
    * [Console](#domino-console)
* [Common Issues](#common-problems)
* [Roadmap](#roadmap)
* [Contributing](#contributing)
* [License](#license)
* [Contact](#authors)
* [Acknowledgements](#acknowledgments)


## DominoVagrant
Primary goal is to use Vagrant to deploy the latest Domino Server in an Linux VM. Vagrant and Role Specific Variables will be passed along, automating installation via the RestAPI interace and Mooneshine or other tools that support CRUD API calls. This uses a Specialized Packer Build that cuts down deployment time:

* **Template:** [Packer](https://app.vagrantup.com/STARTcloud/boxes/debian11-server)
* **Build Source:** [Repo](Notyetavailableforpublicconsumption)

Each Release will be a at the time, stable branch. Recommended to use the latest.

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes, as well as what will power the build process of the VMs at Prominic.NET.

### Prerequisites

You will need some software on your PC or Mac:

```
git
Vagrant
Virtualbox
```

## Installation

To ease deployment, we have a few handy scripts that will utlize a package manager for each OS to get the pre-requisite software for your host OS. This is NOT required, this is to help you ensure you have all the applications that are neccessary to run this VM.

#### Windows
Powershell has a package manager named Chocalatey which is very similar to SNAP, YUM, or other Package manager, We will utilize that to quickly install Virtualbox, Vagrant and Git.

Powershell
```powershell
Set-ExecutionPolicy Bypass -Scope Process -Force; iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))
choco install vagrant
choco install virtualbox
choco install git.install
```

For those that need to run this in a Command Prompt, you can use this:

CMD
```bat
@"%SystemRoot%\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))" && SET "PATH=%PATH%;%ALLUSERSPROFILE%\chocolatey\bin"
choco install vagrant
choco install virtualbox
choco install git.install
```

#### Mac
Just like Windows and other Linux repos, there is a similar package manager for Mac OS X, Homebrew, We will utilize that to install the prequsites. You will likley need to allow unauthenticated applications in the Mac OS X Security Settings, there are reports that Mac OS X Mojave will require some additional work to get running correctly. You do NOT have to use these scripts to get the pre-requisites on your Mac, it is recommened, you simply need to make sure you have the 3 applications installed on your Mac.

```shell
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
brew cask install virtualbox
brew cask install vagrant
brew cask install vagrant-manager
brew install git
```

#### CentOS 7
We will utilize YUM and a few other bash commands to get the Virtualbox, Git,  and Vagrant installed.

YUM
```shell
yum -y install gcc dkms make qt libgomp patch kernel-headers kernel-devel binutils glibc-headers glibc-devel font-forge
cd /etc/yum.repo.d/
wget http://download.virtualbox.org/virtualbox/rpm/rhel/virtualbox.repo
yum install -y VirtualBox-5.1
/sbin/rcvboxdrv setup
yum -y install https://releases.hashicorp.com/vagrant/1.9.6/vagrant_1.9.6_x86_64.rpm
sudo yum install git
```

#### Ubuntu
We will utilize APT to get the Virtualbox, Git,  and Vagrant installed.

APT
```shell
sudo apt-get install virtualbox vagrant git-core -y
```

## Deployment
### Cloning the repo locally

Open up a terminal and perform the following git command in order to save the Project to a local folder:

```shell
git clone https://github.com/DominoVagrant/demo-v2-task-based
```

### Configuring the Environment
Once you have navigated into the projects directory. You will need to modify the Hosts.yml to your specific Environment.

Please set the configuration file with the correct, Network and Memory and CPU settings your host machine will allow, as these may vary from system to system, I cannot predict your Machines CPU and Network Requirements. You will need to make sure that you do not over allocate CPU, and RAM.

##### Networking is setup to create one NAT adapter for Vagrant communications and one Bridged Adapter.
The bridge adapter needs to be specified or it will prompt upon deployment.
Setting dhcp4 to true (ipv6 not yet fully supported try at your own risk) will pull a IP from your local network's DHCP server.

##### Secrets

If you have any sensitive credentials, You will also need to create ```.secrets.yml``` in the root of the project. This is where you can store credentials variables that may contain sensitive data. this will prevent you from uploading them to the repo should you contribute back. Please note that if you remove this from the .gitignore you risk uploading sensitve data.

```
cd demo-v2-task-based
touch .secrets.yml
nano Hosts.yml
```

## Commonly Changed Parameters:

* ip: Use any IP on your Internal Network that is NOT in use by any other machine.
* gateway: This is the IP of your Router
* dhcp4: true/false
* hostname: This is the Hostname of the VM,
* domain: This is the domain to complete the FQDN
* mac: This is your machines Network unique identifier, if you run more than one instance on a network, randonmize this. [Mac Generator](https://www.miniwebtool.com/mac-address-generator/)
* netmask: Set this to the subnet you have your network configured to. This is normally: 255.255.255.0
* name: The Vagrant unique identifier
* cpu: The number of cores you are allocating to this machine. Please beware and do not over allocate. Overallocation may cause instability
* memory: The amount of Memory you are allocating to the machine.  Please beware and do not over allocate. Overallocation may cause instability



### Modifying Roles
The default provisioning engine is ansible-local. This allows us to template our variables into files before deploying and executing the installers.
This allows us to set dynamic usernames, paths, passwords, etc.

#### Domino One-Touch References
In order to make changes to the one touch installer. Modify the template file setup.json.j2 in the /templates folder of the role "domino-config".

You can find more information on the fields and how they correspond to Field Values in Doimino designer here:

[Domino-OneTouch](https://help.hcltechsw.com/domino/12.0.0/admin/inst_usingthedominoserversetupprogram_c.html)

## Source Files

If you have Domino and the installations files in a remote repository.
You can define them in the Hosts.yml under their respective variables.

If you do not have a repository to pull your installation files from.
You can place the archived installers in the ./installers/{{APPLICATION}}/archived directory.
These will be expanded into their respective folders under /vagrant/installers/{{APPLICATION}}/archived.

You will need to supply the Domino installer and optional fix pack files
yourself (eg, Domino_12.0_Linux_English.tar, Domino_1101FP2_Linux.tar).

## Cross Certifying

If you want to access the server from a Notes ID, create a safe ID using the instructions [here](#access-from-notes-client)

**Place your file into the ./safe-id-to-cross-certify folder.**

## Starting the VM
The installation process is estimated to take about 15 - 30 Minutes.

```
vagrant up
```

At this point, you can execute 'vagrant up' in the git checkout directory
to spin up a vm instance, or use the utility scripts
./scripts/vagrant_up.sh, ./scripts/vagrant_up.ps1 to create a log file with the initialization
output in addition to showing on the screen.

Once the system has been provisioned, you can use 'vagrant ssh' to access
it, or again the utility scripts vagrant_ssh.sh/vagrant_ssh.ps1 to create
a log file of the ssh session.

View the contents of the CommandHelp.text for more details.
This file will also be displayed followed each vagrant up operation for
your continued reference.

## Accessing the Domino Server

The Domino server will be started automatically when `vagrant up` completes.

### Domino Console

To access the console, run:

```vagrant ssh -c "screen -r"```
or
```vagrant ssh -c "sudo domino console"```

### Web Interface

The web interface of the server is here: https://yourstaticordhcpip:443/downloads/welcome.html

### Access from Notes Client

If you want to access the server from a Notes Client, you will need to cross-certify your ID.  To do this, first create a safe ID:
1. Open User Security:
	- MacOS: HCL Notes > Security > User Security
	- Windows: File > Security > User Security
2. Select the Your Identity > Your Certificates tab
3. Run Other Actions > Export NotesID Safe ID.  Do not set a password

Copy this ID to `./safe-id-to-cross-certify` and update `safe_notes_id`, and run `vagrant up`.

Then you will need to create a connection document in your local Notes client.
1. File > Open > HCL Notes Application
2. Open names.nsf on your local machine
3. Click `Advanced` in the bottom of the left sidebar
4. Open the Connections view
5. Click New > Server Connection
	1. In the Basic tab, set `Server name` as "demo/Demo" and check the `TCP/IP` checkbox
	2. In the Advanced tab, set the `Destination server address` to "127.0.0.1:1352"
	3. Click `Save & Close`

Then you can open a database on the server like this:
1. File > Open > HCL Notes Application
2. Enter "demo/DEMO" as the server name
3. Select a database (like names.nsf) and click Open

### Domino Default Credentials

* username: Demo Admin
* password: password

## Common Problems

### Error for Headless VirtualBox

If you get an error indicating that VirtualBox could not start in headless mode, open Vagrantfile and uncomment this line

```
     #vb.gui = true
```

## Roadmap

See the [open issues](https://github.com/DominoVagrant/demo-v2-task-based/issues) for a list of proposed features (and known issues).

## Built With
* [Vagrant](https://www.vagrantup.com/) - Portable Development Environment Suite.
* [VirtualBox](https://www.virtualbox.org/wiki/Downloads) - Hypervisor.
* [Ansible](https://www.ansible.com/) - Virtual Manchine Automation Management.

## Contributing

Please read [CONTRIBUTING.md](https://www.prominic.net) for details on our code of conduct, and the process for submitting pull requests to us.

## Authors
* **Joel Anderson** - *Initial work* - [JoelProminic](https://github.com/JoelProminic)
* **Justin Hill** - *Initial work* - [JustinProminic](https://github.com/JustinProminic)
* **Mark Gilbert** - *Refactor* - [MarkProminic](https://github.com/MarkProminic)

See also the list of [contributors](https://github.com/DominoVagrant/demo-v2-task-based/graphs/contributors) who participated in this project.

## License

This project is licensed under the SSLP v3 License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgments

* Hat tip to anyone whose code was used


CHANGELOG.md
# Changelog

## [0.1.23](https://github.com/DominoVagrant/demo-tasks/compare/demo-tasks/v0.1.22...demo-tasks/v0.1.23) (2023-09-20)


### Bug Fixes

* auto-buildandrelease ([7de4114](https://github.com/DominoVagrant/demo-tasks/commit/7de41148ba3ff050988cffbddb813dbb765ac689))
* auto-buildandrelease ([3180731](https://github.com/DominoVagrant/demo-tasks/commit/3180731b74b16d211ab42d841ba11da8dee1eb71))
* auto-buildandrelease ([302d856](https://github.com/DominoVagrant/demo-tasks/commit/302d85694c2a284be5493fb5fe43efdea9518fc4))
* auto-buildandrelease ([9da7d1a](https://github.com/DominoVagrant/demo-tasks/commit/9da7d1a3f2dc29ae1a34375b2edbbd77f9de28f4))
* auto-buildandrelease ([63a2ed6](https://github.com/DominoVagrant/demo-tasks/commit/63a2ed6e10e10423bc3e02b677ba3e54af655c76))
* auto-buildandrelease ([576d8d4](https://github.com/DominoVagrant/demo-tasks/commit/576d8d48b3ce227c0fe386395848edb01f114e1a))
* auto-buildandrelease ([6ce3798](https://github.com/DominoVagrant/demo-tasks/commit/6ce37989e5f8f677453aa40164349fd29a196f30))
* auto-buildandrelease ([3b7e2f0](https://github.com/DominoVagrant/demo-tasks/commit/3b7e2f0ff06f8422db77190939e9c3fbfa32cc83))
* include core provisioner ([e2115f0](https://github.com/DominoVagrant/demo-tasks/commit/e2115f0709b1afe4984a996782dd1cf15bb45b91))
* include core provisioner ([69c04b6](https://github.com/DominoVagrant/demo-tasks/commit/69c04b6dc244ba6f515a75faeb63ff6e6bb4a3b7))
* include core provisioner ([cff5e64](https://github.com/DominoVagrant/demo-tasks/commit/cff5e648b531d42ef055096a3bf21156b1808873))
* include core provisioner ([46181cc](https://github.com/DominoVagrant/demo-tasks/commit/46181cc7380ca0b03d480e899ffb7d7138e6dad1))
* include core provisioner ([b7197b2](https://github.com/DominoVagrant/demo-tasks/commit/b7197b2df5e507eea633053c76d50eb8a50ba519))
* include core provisioner ([2d5e244](https://github.com/DominoVagrant/demo-tasks/commit/2d5e244d6f6fdf0ad2d0b80200b03658cdf3782e))
* include core provisioner ([82304e8](https://github.com/DominoVagrant/demo-tasks/commit/82304e8359a8e1ccc4237a6cfe4a59e431713842))
* include core provisioner ([05448a4](https://github.com/DominoVagrant/demo-tasks/commit/05448a4823df0e795275e82c32f25cf629d91ebf))
* include core provisioner ([d4b0f28](https://github.com/DominoVagrant/demo-tasks/commit/d4b0f28c18b5fc25bb273d930842f46090f69568))
* include core provisioner ([d777a74](https://github.com/DominoVagrant/demo-tasks/commit/d777a74eaf5cb89a89b907e04ef4633244ddd0cb))
* include core provisioner ([c1b7c00](https://github.com/DominoVagrant/demo-tasks/commit/c1b7c008a2e0b9a88cc6fd8fa4a6d904d2a79545))
* update ([898b98d](https://github.com/DominoVagrant/demo-tasks/commit/898b98d24afcf7e1f44c6efb60c9d0f549f578ba))

## [0.1.22](https://github.com/DominoVagrant/demo-tasks/compare/demo-tasks/v0.1.21...demo-tasks/v0.1.22) (2023-09-20)


### Bug Fixes

* auto-buildandrelease ([01cb66e](https://github.com/DominoVagrant/demo-tasks/commit/01cb66e62c9a19119a4faffbb004443432d9ace7))
* auto-buildandrelease ([606ff69](https://github.com/DominoVagrant/demo-tasks/commit/606ff6976c511cec0b12fc5440ea2f0a9c12cca9))
* auto-buildandrelease ([387ae13](https://github.com/DominoVagrant/demo-tasks/commit/387ae1399921726eacf7ac23def1f78f4c57d309))
* auto-buildandrelease ([4a82874](https://github.com/DominoVagrant/demo-tasks/commit/4a82874c7056b0cc39a92d25379695fdd25c1b72))
* auto-buildandrelease ([e9ea2e2](https://github.com/DominoVagrant/demo-tasks/commit/e9ea2e20670ae0610d761fce648754bb1eaeb7d0))
* githubactions ([fc470f5](https://github.com/DominoVagrant/demo-tasks/commit/fc470f5cea8d0e968f4c667abe47b84720050e2c))

## [0.1.20](https://github.com/DominoVagrant/demo-tasks/compare/demo-tasks/v0.1.19...demo-tasks/v0.1.20) (2023-07-05)


### Bug Fixes

* Website Terminal Fix ([f78cb08](https://github.com/DominoVagrant/demo-tasks/commit/f78cb0869fe79fe66a273cf7b8c631c09f1c77b3))

## [0.1.19](https://github.com/DominoVagrant/demo-tasks/compare/demo-tasks/v0.1.18...demo-tasks/v0.1.19) (2023-07-05)


### Bug Fixes

* Debian 14 support ([f5f5c27](https://github.com/DominoVagrant/demo-tasks/commit/f5f5c27e4cac63b5433a34793259d290f96f4466))
* Debian 14 support ([31f32d8](https://github.com/DominoVagrant/demo-tasks/commit/31f32d8baca93ac835f25709c8dc9e35c3f62dba))
* domino versioning ([1c78b41](https://github.com/DominoVagrant/demo-tasks/commit/1c78b41110d05b1161df45be3b02ba06050069f7))
* remove major, minor, fix for domino installer ([bb839cc](https://github.com/DominoVagrant/demo-tasks/commit/bb839cc153d5e123bdc55ea16d63278430689a74))
* update role remove portforwarding ([21a503a](https://github.com/DominoVagrant/demo-tasks/commit/21a503acefc22dc50ceba05db59199688e6ecf5c))

## [0.1.18](https://github.com/DominoVagrant/demo-tasks/compare/demo-tasks/v0.1.17...demo-tasks/v0.1.18) (2023-02-01)


### Bug Fixes

* port-fwrding and roles, prepare new feats ([034f584](https://github.com/DominoVagrant/demo-tasks/commit/034f584a90d2cded4452111aa4e01022b60e35e4))

## [0.1.17](https://github.com/DominoVagrant/demo-tasks/compare/demo-tasks/v0.1.16...demo-tasks/v0.1.17) (2023-01-25)


### Bug Fixes

* remove Browser opening ([20d2e9c](https://github.com/DominoVagrant/demo-tasks/commit/20d2e9c8daf79d385b15ac2a6d29f5852fb71af2))

## [0.1.16](https://github.com/DominoVagrant/demo-tasks/compare/demo-tasks/v0.1.15...demo-tasks/v0.1.16) (2023-01-25)


### Bug Fixes

* ssl self signed enabled ([ee0fb5d](https://github.com/DominoVagrant/demo-tasks/commit/ee0fb5d856236cd7f92578e5d66c7211ddb0cabf))

## [0.1.15](https://github.com/DominoVagrant/demo-tasks/compare/demo-tasks-v0.1.14...demo-tasks/v0.1.15) (2023-01-25)


### Bug Fixes

* change component ([5c26dce](https://github.com/DominoVagrant/demo-tasks/commit/5c26dce9ec7d5ed51fbef06a8c8e586805ad1401))
* Change versioning to Manifest ([f8389a1](https://github.com/DominoVagrant/demo-tasks/commit/f8389a16fe3d91750085a2f791f050f58ef406e5))

demo-tasks/core/Hosts.rb
# coding: utf-8
require File.expand_path("#{File.dirname(__FILE__)}/version.rb")

# This class takes the Hosts.yaml and set's the neccessary variables to run provider specific sequences to boot a VM.
class Hosts
  def Hosts.configure(config, settings)

    ## Load your Secrets file
    secrets = YAML::load(File.read("#{File.dirname(__FILE__)}/.secrets.yml")) if File.exists?("#{File.dirname(__FILE__)}/.secrets.yml")

    # Main loop to configure VM
    settings['hosts'].each_with_index do |host, index|
      provider = host['provider-type']

      if host.has_key?('plugins')
        host['plugins'].each do |plugin|
          unless Vagrant.has_plugin?("#{plugin}")
            system("vagrant plugin install #{plugin}")
            exit system('vagrant', *ARGV)
          end
        end
      end
      config.vm.define "#{host['settings']['server_id']}--#{host['settings']['hostname']}.#{host['settings']['domain']}" do |server|

        #Box Settings -- Used in downloading and packaging Vagrant boxes
        server.vm.box = host['settings']['box']
        server.vm.box_version = host['settings']['box_version']
        server.vm.boot_timeout = host['settings']['setup_wait']
        # Setup SSH and Prevent TTY errors
        server.ssh.username = host['settings']['vagrant_user']
        #server.ssh.password =  host['settings']['vagrant_user_pass']
        server.ssh.private_key_path = host['settings']['vagrant_user_private_key_path']
        server.ssh.insert_key = host['settings']['vagrant_insert_key']
        server.ssh.forward_agent = host['settings']['ssh_forward_agent']
        config.vm.communicator = :ssh
        config.winrm.username = host['settings']['vagrant_user']
        config.winrm.password = host['settings']['vagrant_user_pass']
        config.winrm.timeout = host['settings']['setup_wait']
        config.winrm.retry_delay = 30
        config.winrm.retry_limit = 1000

        ## Networking
        ## Note Do not place two IPs in the same subnet on both nics at the same time, They must be different subnets or on a different network segment(ie VLAN, physical seperation for Linux VMs)
        if host.has_key?('networks') and !host['networks'].empty?
          host['networks'].each_with_index do |network, netindex|
              network['vmac'] = network['mac'].tr(':', '') if host['provider-type'] == 'virtualbox'
              ## Use default route as the Gateway device
              bridge = network['bridge'] if defined?(network_bridge)

              vm_interfaces = %x[VBoxManage list bridgedifs].split("\n")
              interfaces = []
              vm_interfaces.each do |line|
                  interfaces.append(line) if line.start_with?('Name')
                  interfaces.append(line) if line.start_with?('Status')
              end

              pair = ""
              pairs = []
              interfaces.each_with_index do |line, index|
                  pair = line if index %2 ==0 and line.start_with?("Name:")
                  pairs.append(pair.sub("Name:", '').strip) if index %2 !=0 and line.include? "Up"
                  pair = "" if index %2 !=0
              end

              defroute = ""
              if not Vagrant::Util::Platform.windows?
                %x[netstat -rn -f inet].split("\n").each do |line|
                    defroute = line.split("\s") if line.include? "UG"
                end
              else
                defroute = %x[powershell "Get-NetRoute -DestinationPrefix '0.0.0.0/0' | Sort-Object -Property { $_.InterfaceMetric + $_.RouteMetric } | Get-NetAdapter  -InterfaceIndex  { $_.ifIndex } |  foreach { $_.InterfaceDescription }"]
              end

              pairs.each_with_index do |active_interface, index|
                bridge = active_interface if !defroute.nil? and active_interface.start_with?(defroute.to_s) and Vagrant::Util::Platform.windows? and !defined?(network_bridge)
                bridge = active_interface if !defroute[7].nil? and active_interface.start_with?(defroute[7]) and Vagrant::Util::Platform.linux? and !defined?(network_bridge)
                bridge = active_interface if !defroute[3].nil? and active_interface.start_with?(defroute[3]) and Vagrant::Util::Platform.darwin? and !defined?(network_bridge)
              end

              if network['type'] == 'external'
                server.vm.network "public_network",
                  ip: network['address'],
                  dhcp: network['dhcp4'],
                  dhcp6: network['dhcp6'],
                  bridge: bridge,
                  auto_config: network['autoconf'],
                  netmask: network['netmask'],
                  vmac: network['mac'],
                  mac: network['vmac'],
                  gateway: network['gateway'],
                  nictype: network['type'],
                  nic_number: netindex,
                  managed: network['is_control'],
                  vlan: network['vlan']
              end
              if network['type'] == 'host'
                server.vm.network "private_network",
                  ip: network['address'],
                  netmask: network['netmask']
              end
          end
        end

        ##### Disk Configurations #####
        ## https://sleeplessbeastie.eu/2021/05/10/how-to-define-multiple-disks-inside-vagrant-using-virtualbox-provider/
        disks_directory = File.join("./", "disks")
        
        ## Create Disks
        config.trigger.before :up do |trigger|
          if host.has_key?('disks') and !host['disks'].empty?
            trigger.name = "Create disks"
            trigger.ruby do
              unless File.directory?(disks_directory)
                FileUtils.mkdir_p(disks_directory)
              end
              
              host['disks']['additional_disks'].each_with_index do |disks, diskindex|
                local_disk_filename = File.join(disks_directory, "#{disks['volume_name']}.vdi")
                unless File.exist?(local_disk_filename)
                  puts "Creating \"#{disks['volume_name']}\" disk with size \"#{disks['size'].delete('^0-9').to_i * 1024}\""
                  system("VBoxManage createmedium --filename #{local_disk_filename} --size #{disks['size'].delete('^0-9').to_i * 1024} --format VDI")
                end
              end
            end  
          end
        end

        # create storage controller on first run
        if host.has_key?('disks') and !host['disks'].empty?
          unless File.directory?(disks_directory)
            config.vm.provider "virtualbox" do |storage_provider|
              storage_provider.customize ["storagectl", :id, "--name", "Virtual I/O Device SCSI controller", "--add", "virtio-scsi", '--hostiocache', 'off']
            end
          end
        end

        # attach storage devices
        if host.has_key?('disks') and !host['disks'].empty?
          config.vm.provider "virtualbox" do |storage_provider|
            host['disks']['additional_disks'].each_with_index do |disks, diskindex|
              local_disk_filename = File.join(disks_directory, "#{disks['volume_name']}.vdi")
              unless File.exist?(local_disk_filename)
                storage_provider.customize ['storageattach', :id, '--storagectl', "Virtual I/O Device SCSI controller", '--port', disks['port'], '--device', 0, '--type', 'hdd', '--medium', local_disk_filename]
              end
            end
          end
        end

        # cleanup after "destroy" action
        config.trigger.after :destroy do |trigger|
          if host.has_key?('disks') and !host['disks'].empty?
            trigger.name = "Cleanup operation"
            trigger.ruby do
              # the following loop is now obsolete as these files will be removed automatically as machine dependency
              host['disks']['additional_disks'].each_with_index do |disks, diskindex|
                local_disk_filename = File.join(disks_directory, "#{disks['volume_name']}.vdi")
                if File.exist?(local_disk_filename)
                  puts "Deleting \"#{disks['volume_name']}\" disk"
                  system("vboxmanage closemedium disk #{local_disk_filename} --delete")
                end
              end
              if File.exist?(disks_directory)
                FileUtils.rmdir(disks_directory)
              end
            end
          end
        end

        ##### Begin Virtualbox Configurations #####
        server.vm.provider :virtualbox do |vb|
          if host['settings']['memory'] =~ /gb|g|/
            host['settings']['memory']= 1024 * host['settings']['memory'].tr('^0-9', '').to_i
          elsif host['settings']['memory'] =~ /mb|m|/
            host['settings']['memory']= host['settings']['memory'].tr('^0-9', '')
          end
          vb.name = "#{host['settings']['server_id']}--#{host['settings']['hostname']}.#{host['settings']['domain']}"
          vb.gui = host['settings']['show_console']
          vb.customize ['modifyvm', :id, '--ostype', host['settings']['os_type']]
          vb.customize ["modifyvm", :id, "--vrdeport", host['settings']['consoleport']]
          vb.customize ["modifyvm", :id, "--vrdeaddress", host['settings']['consolehost']]
          vb.customize ["modifyvm", :id, "--cpus", host['settings']['vcpus']]
          vb.customize ["modifyvm", :id, "--memory", host['settings']['memory']]
          vb.customize ["modifyvm", :id, "--firmware", 'efi'] if host['settings']['firmware_type'] == 'UEFI'
          vb.customize ['modifyvm', :id, "--vrde", 'on']
          vb.customize ['modifyvm', :id, "--natdnsproxy1", 'off']
          vb.customize ['modifyvm', :id, "--natdnshostresolver1", 'off']
          vb.customize ['modifyvm', :id, "--accelerate3d", 'off']
          vb.customize ['modifyvm', :id, "--vram", '256']

          if host.has_key?('roles') and !host['roles'].empty?
            host['roles'].each do |rolefwds|
              if rolefwds.has_key?('port_forwards') and !rolefwds.empty?
                rolefwds['port_forwards'].each_with_index do |param, index|
                  config.vm.network "forwarded_port", guest: param['guest'], host: param['host'], host_ip: param['ip']
                end
              end
            end
          end

          if host.has_key?('vbox') and !host['vbox'].empty?
            if host['vbox'].has_key?('directives') and !host['vbox']['directives'].empty?
              host['vbox']['directives'].each do |param|              
                vb.customize ['modifyvm', :id, "--#{param['directive']}", param['value']]
              end
            end
          end
        end
        ##### End Virtualbox Configurations #####

        # Register shared folders
        if host.has_key?('folders')
					host['folders'].each do |folder|
						mount_opts = folder['type'] == folder['type'] ? ['actimeo=1'] : []
						server.vm.synced_folder folder['map'], folder ['to'],
						type: folder['type'],
						owner: folder['owner'] ||= host['settings']['vagrant_user'],
						group: folder['group'] ||= host['settings']['vagrant_user'],
						mount_options: mount_opts,
						automount: true,
            rsync__args: folder['args'] ||= ["--verbose", "--archive", "-z", "--copy-links"],
						rsync__chown: folder['chown'] ||= 'false',
            create: folder['create'] ||= 'false',
						rsync__rsync_ownership: folder['rsync_ownership'] ||= 'true',
						disabled: folder['disabled']||= false
					end
				end
        if host.has_key?('provisioning') and !host['provisioning'].nil?
          # Add Branch Files to Vagrant Share on VM Change to Git folders to pull
          if host['provisioning'].has_key?('role') && host['provisioning']['role']['enabled']
            scriptsPath = File.dirname(__FILE__) + '/scripts'
            server.vm.provision 'shell' do |s|
              s.path = scriptsPath + '/add-role.sh'
              s.args = [host['provisioning']['role']['name'], host['provisioning']['role']['git_url'] ]
            end
          end

          # Run the shell provisioners defined in hosts.yml
          if host['provisioning'].has_key?('shell') && host['provisioning']['shell']['enabled']
          host['provisioning']['shell']['scripts'].each do |file|
              server.vm.provision 'shell', path: file
          end
          end

          # Run the Ansible Provisioners -- You can pass Host.yaml variables to Ansible via the Extra_vars variable as noted below.
          ## If Ansible is not available on the host and is installed in the template you are spinning up, use 'ansible-local'
          if host['provisioning'].has_key?('ansible') && host['provisioning']['ansible']['enabled']
          host['provisioning']['ansible']['scripts'].each do |scripts|
            if scripts.has_key?('local')
              scripts['local'].each do |localscript|
                server.vm.provision :ansible_local do |ansible|
                  ansible.playbook = localscript['script']
                  ansible.compatibility_mode = localscript['compatibility_mode'].to_s
                  ansible.install_mode = "pip" if localscript['install_mode'] == "pip"
                  ansible.verbose = localscript['verbose']
                  ansible.config_file = "/vagrant/ansible/ansible.cfg"
                  ansible.extra_vars = {
                    settings: host['settings'],
                    networks: host['networks'],
                    secrets: secrets,
                    role_vars: host['vars'],
                    provision_roles: host['roles'],
                    demo_tasks_version: DemoTasks::VERSION,
                    ansible_winrm_server_cert_validation: "ignore",
                    ansible_ssh_pipelining:localscript['ssh_pipelining'],
                    ansible_python_interpreter:localscript['ansible_python_interpreter']}
                end
              end
            end

            ## If Ansible is available on the host or is not installed in the template you are spinning up, use 'ansible'
            if scripts.has_key?('remote')
              scripts['remote'].each do |remotescript|
                server.vm.provision :ansible do |ansible|
                  ansible.playbook = remotescript['script']
                  ansible.compatibility_mode = remotescript['compatibility_mode'].to_s
                  ansible.verbose = remotescript['verbose']
                  ansible.extra_vars = {
                    settings: host['settings'],
                    networks: host['networks'],
                    secrets: secrets,
                    role_vars: host['vars'],
                    provision_roles: host['roles'],
                    demo_tasks_version: DemoTasks::VERSION,
                    ansible_winrm_server_cert_validation: "ignore",
                    ansible_ssh_pipelining:remotescript['ssh_pipelining'],
                    ansible_python_interpreter:remotescript['ansible_python_interpreter']
                  }
                end
              end
            end
          end
          end

          # Run the Docker-Compose provisioners defined in hosts.yml
          if host['provisioning'].has_key?('docker') && host['provisioning']['docker']['enabled']
          host['provisioning']['docker']['docker_compose'].each do |file|
              server.vm.provision 'docker'
              server.vm.provision :docker_compose, yml: file, run: "always"
          end
          end
        end
      end

      ## Save variables to .vagrant directory
      if host.has_key?('networks') && host['settings']['provider-type'] == 'virtualbox'
        host['networks'].each_with_index do |network, netindex|
          config.trigger.after [:up] do |trigger|
            trigger.ruby do |env,machine|
              puts "This server has been provisioned with DemoTasks Roles v" + DemoTasks::VERSION
              puts "https://github.com/DominoVagrant/demo-tasks/releases/tag/v" + DemoTasks::VERSION
              ipaddress = network['address']
              system("vagrant ssh -c 'cat /vagrant/completed/ipaddress.yml' > .vagrant/provisioned-briged-ip.txt")
              ipaddress = File.readlines(".vagrant/provisioned-briged-ip.txt").join("") if network['dhcp4']
              open_url = "https://" + ipaddress + ":443/welcome.html"
              system("echo '" + open_url + "' > .vagrant/done.txt")
            end
          end
        end
      end
    end
  end
end
